pkg/endpoint/bpf.go

```diff
@@ -1072,58 +1064,66 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
 	return nil
 }
 
-// syncPolicyMapWith updates the bpf policy map state based on the
-// difference between the given 'realized' and desired policy state without
+// syncRealizedPolicyMap updates the bpf policy map state based on the
+// difference between the realized and desired policy state without
 // dumping the bpf policy map.
-// Changes are synced to endpoint's realized policy mapstate, 'realized' is
-// not modified.
-func (e *Endpoint) syncPolicyMapWith(realized policy.MapState, withDiffs bool) (diffCount int, diffs []policy.MapChange, err error) {
+// Only called when desired and realized policies are not the same.
+func (e *Endpoint) syncRealizedPolicyMap() error {
```

Comment:
Is there a reason this is different that `syncPolicyMapWith(e.realized)`?

The naming and functionality between these two functions is a bit confusing; they take a purported base state ("realized"), a new state ("desired") and apply the diffs to BPF but not the base state.

My suggestion: create a function, `e.applyPolicyMap(realized, desired, withDiffs)` that does:

1. if `realized` is nil, then dumps the current BPF map state
2. determines delta between desired and realized
3. applies that delta, gathering diffs for debugging if interesting.

This should allow us to unify all uses of this.

Also, the revert function in `e.setDesiredPolicy()` line 365 may be buggy now -- what should it be doing?

Reply:
The 'realized' map is of a different type between these functions. It is the bpf policy map that is being synced. As an immediate remedy to limit confusion I've renamed `syncRealizedPolicyMap` as `syncPolicyMap` (it is not the agent's notion of the realized map that is being synced, but the bpf policy map) and added docblocks.

Reply:
The general strategy on revert is to use the prior `e.realizedPolicy` as the "last known good policy" (it is the state of the policy map before this regeneration started). `e.desiredPolicy` is set to `e.realizedPolicy` first, then a fresh dump of the policy map is taken, and the bpf policy map is synced using those two. After this the bpf policy map should have the entries in `e.desiredPolicy`.